source/extensions/filters/http/common/compressor/compressor.h

```diff
@@ -0,0 +1,197 @@
+namespace Common {
+namespace Compressors {
+
+/**
+ * All compressor filter stats. @see stats_macros.h
+ * "total_uncompressed_bytes" only includes bytes from requests that were marked for compression.
+ * If the request was not marked for compression, the filter increments "not_compressed", but does
+ * not add to "total_uncompressed_bytes". This way, the user can measure the memory performance of
+ * the compression.
+ * "header_gzip" is specific to the gzip filter and is deprecated since it duplicates
+ * "header_compressor_used".
+ */
+// clang-format off
+#define ALL_COMPRESSOR_STATS(COUNTER)    \
+  COUNTER(compressed)                    \
+  COUNTER(not_compressed)                \
+  COUNTER(no_accept_header)              \
+  COUNTER(header_identity)               \
+  COUNTER(header_gzip)                   \
+  COUNTER(header_compressor_used)        \
+  COUNTER(header_compressor_overshadowed)\
+  COUNTER(header_wildcard)               \
+  COUNTER(header_not_valid)              \
+  COUNTER(total_uncompressed_bytes)      \
+  COUNTER(total_compressed_bytes)        \
+  COUNTER(content_length_too_small)      \
+  COUNTER(not_compressed_etag)           \
+// clang-format on
+
+/**
+ * Struct definition for compressor stats. @see stats_macros.h
+ */
+struct CompressorStats {
+  ALL_COMPRESSOR_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+class CompressorFilterConfig {
+
+public:
+  CompressorFilterConfig() = delete;
+
+  virtual const std::string featureName() const PURE;
+  virtual std::unique_ptr<Compressor::Compressor> makeCompressor() PURE;
+
+  Runtime::Loader& runtime() { return runtime_; }
+  CompressorStats& stats() { return stats_; }
+  const StringUtil::CaseUnorderedSet& contentTypeValues() const { return content_type_values_; }
+  bool disableOnEtagHeader() const { return disable_on_etag_header_; }
+  bool removeAcceptEncodingHeader() const { return remove_accept_encoding_header_; }
+  uint32_t minimumLength() const { return content_length_; }
+  const std::string contentEncoding() const { return content_encoding_; };
+  const std::map<std::string, uint32_t> registeredCompressors() const;
+
+  virtual ~CompressorFilterConfig();
+
+protected:
+  CompressorFilterConfig(const envoy::config::filter::http::compressor::v2::Compressor& compressor,
+                         const std::string& stats_prefix, Stats::Scope& scope, Runtime::Loader& runtime,
+                         const std::string& content_encoding);
+
+private:
+  static StringUtil::CaseUnorderedSet
+  contentTypeSet(const Protobuf::RepeatedPtrField<std::string>& types);
+
+  static uint32_t contentLengthUint(Protobuf::uint32 length);
+
+  static CompressorStats generateStats(const std::string& prefix, Stats::Scope& scope) {
+    return CompressorStats{ALL_COMPRESSOR_STATS(POOL_COUNTER_PREFIX(scope, prefix))};
+  }
+
+  struct CompressorRegistry {
+    Thread::MutexBasicLockable mutex_;
+    std::map<CompressorFilterConfig*, uint32_t> compressors_ ABSL_GUARDED_BY(mutex_);
```

Comment:
you might as well make this an absl::flat_hash_map but it's a secondary concern to the mutex.